### PR TITLE
Issue 7 Leaderboard Page

### LINKED ISSUE #7: Admin Portal Wireframes - Leaderboard (Student Rego)
## Basic Information
**[Give a brief summary of what the issue is about]**
TASK:  Make a leaderboard of all the students participating in the competition, this will show their different stats for the admin to see and edit teams and individual members.

- The leaderboard table will not be updated in real time but rather will only be filtered once all of the results for the competition have returned.

- the tables columns will consist of 
**{ student_name: string ,school , school_email, username, password, individual_score, team_score, status }** 
(when the db is developed these will be subject to change but it would be good to make a clear image of what we want for the table)

- should be able to add delete and save entries of information of students, each row represents a unique student entry 

- details of the participants and the competitions results should be able to later be extracted into a csv file for record keeping 

Screenshot 2024-11-25 at 3 15 14 PM

### PR DESCRIPTION
**Pull request by loklokyx**
## Change Summary
server:
- initial `app` directory, including migrations folder and base files for the route `/api/app/leaderboard/`

client:
- created basic UI with a simple `table`, `dropdown`, and `search`
- implemented search functionality with dropdown and input options
- added AJAX to retrieve BE sample data and search functionality in Leaderboard page `/admin/leaderboard`

### Change Form
**Fill this up (NA if not available). If a certain criteria is not met, can you please give a reason.**

- [Y] The pull request title has an issue number
- [Y] The change works by "Smoke testing" or quick testing
- [NA] The change has tests
- [NA] The change has documentation

## Other Information
server:
- temporarily use sample data before model creation for serialization
- permission validation is currently not implemented

client:
- pending layout implementation to merge Leaderboard body content

# Related issue

- Resolve #7

### CODE REVIEW

**Comment by loklokyx** [comment]
image
This is the current UI of the Table and Search Form

**Review by yunho7687 — CHANGES_REQUESTED** [review]
Amazing works! 
Need to mind that we don't need a new app named "app". there's a app named app in penni repo cuz people misunderstood the objective of the django server, which should be an api server and had already been created as "api". you can take this repo as ref: https://github.com/codersforcauses/csf/tree/main/server

**Comment by yunho7687** [comment]
You can use code below to create a sub-app in Django:
```bash
cd api
django-admin startapp sub-app-name-such-as-auth
```
and don't forget to install the app in `INSTALLED_APPS`.    

Here's a ref to that
https://stackoverflow.com/questions/2862084/how-do-i-create-sub-applications-in-django

**Comment by loklokyx** [comment]
Hi Team,

With the CFC standard, I’ve added JSDoc and pydoc comments to the manager module for better clarity and documentation. Please help review the changes and let me know if any adjustments are needed. 

Thanks!

**Review by Julie-Salazar — APPROVED** [review]
Awesome work lgtm, you do have an empty migrations folder but it is understandable given you have set this up, really appreciate the comment documentation 🙌 I'm good to approve this for now, @yunho7687 might have notes tho

File: client/src/hooks/use-fetch-data.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+ */
+interface QueryConfig<TData, TError = AxiosError>
+  extends Omit<UseQueryOptions<TData, TError>, "queryKey" | "queryFn"> {
+  queryKey: any[];
+  endpoint: string;
+  params?: Record<string, any>;
+}
+
+/**
+ * A custom hook that fetches data from an API using the `useQuery` hook from `react-query`.
```
Comment (yunho7687) [inline client/src/hooks/use-fetch-data.ts:25]: I think this is an excellent example of how to write documentation, and it's worth us all following as a standard.

**Review by yunho7687 — COMMENTED** [review]
**Legend work!** 
I’ve decided to hold off on this PR for now until we have a discussion about it in the next session.

**Review by yunho7687 — APPROVED** [review]

**Pull request merged**